tools/teamcity-diff-check/main.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+/*
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+// This file is controlled by MMv1, any changes made here will be overwritten
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"reflect"
+	"regexp"
+)
+
+var serviceFile = flag.String("service_file", "services_ga.kt", "kotlin service file to be parsed")
+var provider = flag.String("provider", "google", "Specify which provider to run diff_check on")
+
+func main() {
+	flag.Parse()
+
+	servicesPath := fmt.Sprintf("../../provider/%s/services/", *provider)
+	cmd := exec.Command("go", "list", "./...")
+	cmd.Dir = servicesPath
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		fmt.Println(err.Error())
+		return
+	}
+
+	// Regex pattern captures "services" from `go list ../../provider/{{*provider}}/services/..`
+	pattern := regexp.MustCompile(`github\.com\/hashicorp\/terraform-provider-(google|google-beta)\/(google|google-beta)\/services\/(?P<service>\w+)`)
+
+	template := []byte("$service")
+	dst := []byte{}
+
+	googleServices := []string{}
+
+	// For each match of the regex in the content.
+	for _, submatches := range pattern.FindAllSubmatchIndex(stdout, -1) {
+		service := pattern.Expand(dst, template, stdout, submatches)
+		googleServices = append(googleServices, string(service))
+	}
+
+	////////////////////////////////////////////////////////////////////////////////
+	test := exec.Command("ls")
+	test.Dir = "../../provider"
+	root, _ := test.Output()
+	fmt.Println(string(root))
+
+	test.Dir = "../../provider/.teamcity"
+	root, _ = test.Output()
+	fmt.Println(string(root))
+
+	test.Dir = "../../provider/.teamcity/components"
+	root, _ = test.Output()
+	fmt.Println(string(root))
+
+	f, err := os.Open(fmt.Sprintf("../../provider/.teamcity/components/inputs/%s", *serviceFile))
```

Comment:
The confusion about the GHA working with GA but not Beta is because this GHA should validate the service files present in the magic-modules repo, not files copied into the downstream provider repos. I think this change is needed, but you'll need to test it:

```suggestion
	f, err := os.Open(fmt.Sprintf("../mmv1/third_party/terraform/.teamcity/components/inputs/%s", *serviceFile))
```

Only the GA version of the provider contains a .teamcity folder, which is why the GHA currently fails when it's attempting to check the services_beta.kt file in the Beta provider code base. Only the GA provider having a .teamcity folder is intentional.

The GHAs need to download the GA and Beta artefacts to perform the `go list` step in your script and determine which service packages are present, but otherwise you shouldn't need to interact with the downstream providers' files.